### PR TITLE
Collision handler

### PR DESCRIPTION
**Pull request by BenjixD**

### CODE REVIEW

File: Assets/Scripts/Unit/Action/Move/Frames/MoveFrame.cs [hunk 1]
```diff
@@ -3,31 +3,47 @@
 using System.Collections.Generic;
 
 public class MoveFrame : Frame {
-	public override bool Execute(Unit unit, Direction dir, Board board) {
-		Direction absoluteDir = (Direction)(((int)relativeDir + (int)dir) % 4);
-		Vector2 next = nextCoord(unit.tile.coordinate, absoluteDir);
+	public override bool CanExecute(Unit unit, Direction dir, Board board) {
+		Vector2 next = GetRelativeDisplacement(unit, dir, board);
+		return board.CheckCoord(next);
```
Comment (leo-jh-li) [inline Assets/Scripts/Unit/Action/Move/Frames/MoveFrame.cs]: This checks if the relative displacement is a valid coord when it seems like it should be checking the absolute displacement

File: Assets/Scripts/Game/Phase/Timeline/Timeline.cs [hunk 2]
```diff
@@ -36,18 +34,26 @@ public class Timeline : Phase {
...
-				ExecuteUnitCommand(unit, board);
+			Queue<Unit> actions = FilterCanExecute(actionDict[type], board);
+			HashSet<Unit> toExecute = UnitsToExecute(actions);
+			Dictionary<Unit, UnitDisplacement> displacements = GetDisplacements(players, actions, board);
+			List<SimulatedDisplacement> result = SimulateDisplacements(displacements, board);
+
+			foreach(SimulatedDisplacement sim in result) {
+				Unit unit = sim.displacement.unit;
+				if(toExecute.Contains(unit)) {
+					Command command = unit.plan.Dequeue();
```
Comment (leo-jh-li) [inline Assets/Scripts/Game/Phase/Timeline/Timeline.cs]: Units that are filtered out by FilterCanExecute don't have their next command dequeued, and thus get stuck trying to perform their next command until it becomes valid. e.g. if a unit is at 0,0 and their plan is move up once (an invalid move) followed by valid moves (moving right, attacking, etc.) they won't do anything. Not sure if this is intended behaviour.

File: Assets/Scripts/Game/Phase/Timeline/SimulatedDisplacement.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			current = new Vector2(Mathf.Round(current.x), Mathf.Round(current.y));
+		} else if(displacement.type == UnitDisplacementType.ABSOLUTE){
+			current = displacement.GetStartCoordinate();
+		}
+		return current;
+	}
+
+	public void DisplaceUnit(Board board) {
+		Unit unit = displacement.unit;
+		unit.tile = board.GetTile(new Vector2(Mathf.Round(current.x), Mathf.Round(current.y)));
```
[the remark this answers is not in the document]
Comment (BenjixD) [reply Assets/Scripts/Game/Phase/Timeline/SimulatedDisplacement.cs]: Made added a new method Board.MoveUnit(SimulatedDisplacement sim) which moves the unit to the target tile given the simulation.

File: Assets/Scripts/Game/Phase/Timeline/Timeline.cs (hunk 2 above)
Comment (BenjixD) [reply Assets/Scripts/Game/Phase/Timeline/Timeline.cs]: No this is not intended behaviour

Comment (BenjixD) [reply Assets/Scripts/Game/Phase/Timeline/Timeline.cs]: I've decided to move the filterExecuteAction check to after the simulation. The original intent of the filter is supposed to stop frames from executing effects if the movement didn't work or vice versa.

File: Assets/Scripts/Unit/Action/Move/Frames/MoveFrame.cs (hunk 1 above)
Comment (BenjixD) [reply Assets/Scripts/Unit/Action/Move/Frames/MoveFrame.cs]: In this case there should be no difference between RelativeDisplacement and AbsoluteDisplacement.

Comment (leo-jh-li) [reply Assets/Scripts/Unit/Action/Move/Frames/MoveFrame.cs]: Scratch that, I mixed up the terminology. What I meant to say is that CheckCoord is checking RelativeDisplacements but it should be checking the resultant destinations. e.g. if the unit was trying to go up, next would be (-1, 0) which would always fail CheckCoord. I think what you want to pass to CheckCoord is something like unit.tile.coordinate + next

Comment (BenjixD) [reply Assets/Scripts/Unit/Action/Move/Frames/MoveFrame.cs]: Oh you are right - yeah this is only checking the relative displacement but not the actual coordinate. I'll make that fix.

**Pull request merged**